discovery/ecs/client/cache.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2016 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ecs"
+)
+
+const (
+	serviceTaskStatusInactive = "INACTIVE"
+)
+
+// awsCache is a temporal store to map between ids on AWS API objects
+type awsCache struct {
+	// Store indexes
+	clusters   map[string]*ecs.Cluster           // cluster cache
+	cInstances map[string]*ecs.ContainerInstance // container instance cache
+	tasks      map[string]*ecs.Task              // task cache
+	instances  map[string]*ec2.Instance          // instance cache
+	services   map[string]*ecs.Service           // service cache
+	taskDefs   map[string]*ecs.TaskDefinition    // task definition cache
+
+	// mutexes
+	clustersMutex   sync.Mutex
+	cInstancesMutex sync.Mutex
+	tasksMutex      sync.Mutex
+	instancesMutex  sync.Mutex
+	servicesMutex   sync.Mutex
+	taskDefsMutex   sync.Mutex
+}
+
+func newAWSCache() *awsCache {
+	a := &awsCache{
+		clusters:   map[string]*ecs.Cluster{},
+		cInstances: map[string]*ecs.ContainerInstance{},
+		tasks:      map[string]*ecs.Task{},
+		instances:  map[string]*ec2.Instance{},
+		services:   map[string]*ecs.Service{},
+		taskDefs:   map[string]*ecs.TaskDefinition{},
+
+		clustersMutex:   sync.Mutex{},
+		cInstancesMutex: sync.Mutex{},
+		tasksMutex:      sync.Mutex{},
+		instancesMutex:  sync.Mutex{},
+		servicesMutex:   sync.Mutex{},
+		taskDefsMutex:   sync.Mutex{},
```

Comment:
Its a bad practice to initialize explicitly the mutexes? I thought that the same way it's not a bad practice to do `a := 0` instead of `var a` this neither.

I can't guarantee, I guess that some of them can be removed but I need to check them